src/raw/mod.rs

```diff
@@ -137,23 +138,46 @@ fn h1(hash: u64) -> usize {
     hash as usize
 }
 
-// Constant for h2 function that grabing the top 7 bits of the hash.
+// Constant for h2 function that grabing the top 8 bits of the hash.
 const MIN_HASH_LEN: usize = if mem::size_of::<usize>() < mem::size_of::<u64>() {
     mem::size_of::<usize>()
 } else {
     mem::size_of::<u64>()
 };
 
-/// Secondary hash function, saved in the low 7 bits of the control byte.
+/// Secondary hash function, saved in the control byte.
 #[inline]
 #[allow(clippy::cast_possible_truncation)]
 fn h2(hash: u64) -> u8 {
-    // Grab the top 7 bits of the hash. While the hash is normally a full 64-bit
+    const fn compute_control() -> [u8; 256] {
+        let mut result = [0; 256];
+
+        let mut i = 0;
+
+        while i < 256 {
+            result[i] = i as u8;
+
+            i += 1;
+        }
+
+        //  Avoid overlap with special values.
+        result[EMPTY as usize] += 8;
+        result[DELETED as usize] += 8;
+
+        result
+    }
+
+    #[rustfmt::skip]
+    const CONTROL: [u8; 256] = compute_control();
+
+    // Grab the top 8 bits of the hash. While the hash is normally a full 64-bit
     // value, some hash functions (such as FxHash) produce a usize result
     // instead, which means that the top 32 bits are 0 on 32-bit platforms.
     // So we use MIN_HASH_LEN constant to handle this.
-    let top7 = hash >> (MIN_HASH_LEN * 8 - 7);
-    (top7 & 0x7f) as u8 // truncation
+    let top8 = hash >> (MIN_HASH_LEN * 8 - 7);
+
+    // Lookup matching control byte, avoid overlap with special control.
+    CONTROL[top8 as usize]
```

Comment:
On a **_x64_** system, the expression `let top8 = hash >> (MIN_HASH_LEN * 8 - 7)` is a **_truncation_** because you are shifting left by **_57 bits_**. You can look here: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5f549c43d8bbcfba88053292b4a06887). That is, on a **x64** system, the expression `top7 & 0x7f` is completely removed by the compiler. It (`top7 & 0x7f`) only makes sense on 32-bit platforms.

Due to the fact that you are doing benchmarks most likely on an **x64** system, you are seeing a slowdown, since you simply left using the same **7 bits**, **_while adding new logic._**

And maybe I was thinking that if we use the numbers **255** and **254** as **EMPTY** and **DELETED**, and instead of a constant array (which requires allocation), implement the following logic:
```rust
const EMPTY: u8 = 0b1111_1111;   // 255
const DELETED: u8 = 0b1111_1110; // 254

#[no_mangle]
pub fn h2(hash: u64) -> u8 {
    let bit = hash as u8;
    bit >> ((bit > 253) as u8)
}
```
It seems the resulting asm is not so bad [godbolt](https://godbolt.org/z/d3K8hv36h):
```asm
h2:
        mov     rax, rdi
        cmp     al, -2
        setae   cl
        shr     al, cl
        ret
```